src/linear_algebra/bilinear_form.lean

```diff
@@ -0,0 +1,226 @@
+/-
+Copyright (c) 2018 Andreas Swerdlow. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Andreas Swerdlow
+-/
+
+import linear_algebra.tensor_product
+
+universes u v
+
+structure bilin_form (α : Type u) (β : Type v) [ring α] [add_comm_group β] [module α β] :=
+(bilin : β → β → α)
+(bilin_add_left : ∀ (x y z : β), bilin (x + y) z = bilin x z + bilin y z)
+(bilin_smul_left : ∀ (a : α) (x y : β), bilin (a • x) y = a * (bilin x y))
+(bilin_add_right : ∀ (x y z : β), bilin x (y + z) = bilin x y + bilin x z)
+(bilin_smul_right : ∀ (a : α) (x y : β), bilin x (a • y) = a * (bilin x y))
+
+namespace bilin_form
+ 
+open bilin_form
+
+variables {α : Type u} {β : Type v} [ring α] [add_comm_group β] [module α β] {B : bilin_form α β}
+
+instance module : module α β := by apply_instance
+
+lemma zero_bilin (x : β) :
+B.bilin 0 x = 0 := by rw [←@zero_smul α _ _ _ _ (0 : β), bilin_smul_left, zero_mul]
```

Comment:
Can you make a coercion to functions to create nicer notation?

Reply:
Is there a way of doing this without having to rewrite all the fields in bilin_form as lemmas, so lean can use them with the new notation?

Reply:
Usually not, unfortunately. Setting up coercions and notations often have some overhead for the library author.